File: aether_messenger/windows/AetherQMessenger/Data/MessengerRepository.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using AetherQMessenger.Crypto;
using AetherQMessenger.Ffi;
using AetherQMessenger.Mqtt;

namespace AetherQMessenger.Data;

/// <summary>
/// MQTT transport, AUTH-AKEM handshake, Compress-KEM ratchet va mahalliy bazani
/// bir joyga bog'laydigan yuqori darajadagi fasad (Android'dagi
/// <c>MessengerRepository.kt</c>ning C# ekvivalenti).
/// </summary>
public sealed class MessengerRepository
{
    private readonly Identity _identity;
    private readonly AppDatabase _db;
    private readonly SessionManager _sessionManager;
    private readonly MqttClientWrapper _mqttClient;

    /// <summary>Kontakt yoki xabar ro'yxati o'zgarganda ishga tushadi (istalgan thread'dan).</summary>
    public event Action? DataChanged;

    public MessengerRepository(Identity identity, AppDatabase db, SessionManager sessionManager, MqttClientWrapper mqttClient)
    {
        _identity = identity;
        _db = db;
        _sessionManager = sessionManager;
        _mqttClient = mqttClient;
    }

    public List<ContactRecord> GetContacts() => _db.GetAllContacts();

    public List<MessageRecord> GetMessages(string contactUserId) => _db.GetMessagesForContact(contactUserId);

    public void AddContact(ContactCard card, string displayName)
    {
        _db.UpsertContact(new ContactRecord
        {
            UserId = card.UserId,
            DisplayName = displayName,
            EdPkHex = card.EdPk,
            XPkHex = card.XPk,
            MlkemPkHex = card.MlkemPk,
            CreatedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        });
        DataChanged?.Invoke();
    }

    public Task ConnectMqttAsync() => _mqttClient.ConnectAndSubscribeAsync(HandleIncoming);

    public async Task SendMessageAsync(ContactRecord contact, string text)
    {
        await EnsureSessionAsync(contact);
        var session = _sessionManager.SessionFor(contact.UserId) ?? throw new InvalidOperationException("Seans o'rnatilmadi");

        var step = session.Sender.NextFrame(Convert.FromHexString(contact.XPkHex));
        var recordFrame = AetherCrypto.SealRecord(step.SharedSecret, session.SessionId, Encoding.UTF8.GetBytes(text));
        var envelope = MessageEnvelope.EncodeMessage(step.Frame, recordFrame);
        await _mqttClient.PublishToAsync(contact.UserId, envelope);

        _db.InsertMessage(new MessageRecord
        {
            ContactUserId = contact.UserId,
            IsOutgoing = true,
            Plaintext = text,
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        });
        DataChanged?.Invoke();
    }

    /// <summary>Kontakt bilan seans yo'q bo'lsa AUTH-AKEM handshake bajarib MQTT orqali yuboradi.</summary>
    private async Task EnsureSessionAsync(ContactRecord contact)
    {
        if (_sessionManager.HasSession(contact.UserId))
        {
            return;
        }

        var encaps = AetherCrypto.AkemEncapsulate(
            Convert.FromHexString(contact.MlkemPkHex),
            Convert.FromHexString(contact.XPkHex),
            _identity.EdSk,
            _identity.EdPk,
            Convert.FromHexString(contact.EdPkHex));

        _sessionManager.EstablishSession(contact.UserId, encaps.SharedSecret);

        var envelope = MessageEnvelope.EncodeHandshake(Convert.FromHexString(_identity.UserId), encaps);
        await _mqttClient.PublishToAsync(contact.UserId, envelope);
    }

    private void HandleIncoming(byte[] payload)
    {
        switch (MessageEnvelope.Decode(payload))
        {
            case InboundEnvelope.Handshake handshake:
                HandleHandshake(handshake);
                break;
            case InboundEnvelope.RatchetMessage message:
                HandleRatchetMessage(message);
                break;
        }
    }

    private void HandleHandshake(InboundEnvelope.Handshake envelope)
    {
        var senderUserId = Convert.ToHexString(envelope.SenderUserId).ToLowerInvariant();
        // Noma'lum yuboruvchi (oldindan QR/matn orqali qo'shilmagan) e'tiborsiz qoldiriladi.
        var contact = _db.FindContact(senderUserId);
        if (contact == null)
        {
            return;
        }

        var sharedSecret = AetherCrypto.AkemDecapsulate(
            _identity.MlkemSk,
            _identity.XSk,
            _identity.XPk,
            _identity.ZRecip,
            Convert.FromHexString(contact.EdPkHex),
            _identity.EdPk,
            envelope.CtMlkem,
            envelope.CtX25519,
            envelope.Signature);

        _sessionManager.EstablishSession(contact.UserId, sharedSecret);
    }

    private void HandleRatchetMessage(InboundEnvelope.RatchetMessage envelope)
    {
        var sessionIdBytes = envelope.CompressFrame[1..9];
        var contactUserId = _sessionManager.ContactForSessionId(sessionIdBytes);
        if (contactUserId == null)
        {
            return;
        }

        var session = _sessionManager.SessionFor(contactUserId);
        if (session == null)
        {
            return;
        }

        var stepSecret = session.Recipient.AcceptFrame(envelope.CompressFrame, _identity.XSk);
        var plaintext = AetherCrypto.OpenRecord(stepSecret, session.SessionId, envelope.RecordFrame);

        _db.InsertMessage(new MessageRecord
        {
            ContactUserId = contactUserId,
            IsOutgoing = false,
            Plaintext = Encoding.UTF8.GetString(plaintext),
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        });
        DataChanged?.Invoke();
    }
}

File: aether_messenger/android/app/src/main/java/com/aetherq/messenger/data/MessengerRepository.kt
package com.aetherq.messenger.data

import com.aetherq.messenger.crypto.AetherCrypto
import com.aetherq.messenger.crypto.ContactCard
import com.aetherq.messenger.crypto.Identity
import com.aetherq.messenger.crypto.hexToBytes
import com.aetherq.messenger.crypto.toHex
import com.aetherq.messenger.data.local.dao.ContactDao
import com.aetherq.messenger.data.local.dao.MessageDao
import com.aetherq.messenger.data.local.entity.ContactEntity
import com.aetherq.messenger.data.local.entity.MessageEntity
import com.aetherq.messenger.mqtt.InboundEnvelope
import com.aetherq.messenger.mqtt.MessageEnvelope
import com.aetherq.messenger.mqtt.MqttClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

/**
 * MQTT transport, AUTH-AKEM handshake, Compress-KEM ratchet va mahalliy bazani
 * bir joyga bog'laydigan yuqori darajadagi fasad — Compose ViewModel'lar shu orqali
 * ishlaydi, JNI/protokol tafsilotlarini bilishlari shart emas.
 */
class MessengerRepository(
    private val identity: Identity,
    private val contactDao: ContactDao,
    private val messageDao: MessageDao,
    private val sessionManager: SessionManager,
    private val mqttClient: MqttClient,
) {
    // HiveMQ callback'lari background thread'da chaqiriladi — suspend DAO/kripto
    // chaqiruvlarini shu yerdan ishga tushirish uchun alohida scope kerak.
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    fun observeContacts(): Flow<List<ContactEntity>> = contactDao.observeAll()

    fun observeMessages(contactUserId: String): Flow<List<MessageEntity>> = messageDao.observeForContact(contactUserId)

    suspend fun addContact(card: ContactCard, displayName: String) {
        contactDao.upsert(
            ContactEntity(
                userId = card.userId,
                displayName = displayName,
                edPkHex = card.edPk,
                xPkHex = card.xPk,
                mlkemPkHex = card.mlkemPk,
                createdAt = System.currentTimeMillis(),
            ),
        )
    }

    suspend fun connectMqtt() {
        mqttClient.connectAndSubscribe { payload -> scope.launch { handleIncoming(payload) } }
    }

    /** Kontakt bilan seans yo'q bo'lsa AUTH-AKEM handshake bajarib MQTT orqali yuboradi. */
    private suspend fun ensureSession(contact: ContactEntity) {
        if (sessionManager.hasSession(contact.userId)) return

        val encaps =
            AetherCrypto.akemEncapsulate(
                pkRecipMlkem = hexToBytes(contact.mlkemPkHex),
                pkRecipX25519 = hexToBytes(contact.xPkHex),
                skSenderSign = identity.edSk,
                pkSenderSign = identity.edPk,
                pkRecipSign = hexToBytes(contact.edPkHex),
            )
        sessionManager.establishSession(contact.userId, encaps.sharedSecret)

        val envelope = MessageEnvelope.encodeHandshake(hexToBytes(identity.userId), encaps)
        mqttClient.publishTo(contact.userId, envelope)
    }

    suspend fun sendMessage(contact: ContactEntity, text: String) {
        ensureSession(contact)
        val session = sessionManager.sessionFor(contact.userId) ?: error("Seans o'rnatilmadi")

        val step = session.sender.nextFrame(hexToBytes(contact.xPkHex))
        val recordFrame = AetherCrypto.sealRecord(step.sharedSecret, session.sessionId, text.toByteArray(Charsets.UTF_8))
        val envelope = MessageEnvelope.encodeMessage(step.frame, recordFrame)
        mqttClient.publishTo(contact.userId, envelope)

        messageDao.insert(
            MessageEntity(
                contactUserId = contact.userId,
                isOutgoing = true,
                plaintext = text,
                timestamp = System.currentTimeMillis(),
            ),
        )
    }

    private suspend fun handleIncoming(payload: ByteArray) {
        when (val envelope = MessageEnvelope.decode(payload)) {
            is InboundEnvelope.Handshake -> handleHandshake(envelope)
            is InboundEnvelope.RatchetMessage -> handleRatchetMessage(envelope)
            null -> Unit
        }
    }

    private suspend fun handleHandshake(envelope: InboundEnvelope.Handshake) {
        val senderUserId = envelope.senderUserId.toHex()
        // Noma'lum yuboruvchi (QR orqali oldindan qo'shilmagan) e'tiborsiz qoldiriladi —
        // markazlashgan directory yo'q, faqat oldindan almashilgan kalitlarga ishoniladi.
        val contact = contactDao.findByUserId(senderUserId) ?: return

        val sharedSecret =
            AetherCrypto.akemDecapsulate(
                skRecipMlkem = identity.mlkemSk,
                skRecipX25519 = identity.xSk,
                pkRecipX25519 = identity.xPk,
                zRecip = identity.zRecip,
                pkSenderSign = hexToBytes(contact.edPkHex),
                pkRecipSign = identity.edPk,
                ctMlkem = envelope.ctMlkem,
                ctX25519 = envelope.ctX25519,
                signature = envelope.signature,
            )
        sessionManager.establishSession(contact.userId, sharedSecret)
    }

    private suspend fun handleRatchetMessage(envelope: InboundEnvelope.RatchetMessage) {
        val sessionIdBytes = envelope.compressFrame.copyOfRange(1, 9)
        val contactUserId = sessionManager.contactForSessionId(sessionIdBytes) ?: return
        val session = sessionManager.sessionFor(contactUserId) ?: return

        val stepSecret = session.recipient.acceptFrame(envelope.compressFrame, identity.xSk)
        val plaintext = AetherCrypto.openRecord(stepSecret, session.sessionId, envelope.recordFrame)

        messageDao.insert(
            MessageEntity(
                contactUserId = contactUserId,
                isOutgoing = false,
                plaintext = String(plaintext, Charsets.UTF_8),
                timestamp = System.currentTimeMillis(),
            ),
        )
    }
}
